connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSourceTask.java

```diff
@@ -241,17 +290,48 @@ private boolean sendRecords() {
         recordBatch(toSend.size());
         final SourceRecordWriteCounter counter = new SourceRecordWriteCounter(toSend.size(), sourceTaskMetricsGroup);
         for (final SourceRecord preTransformRecord : toSend) {
-            final SourceRecord record = transformationChain.apply(preTransformRecord);
+            processingContext.sourceRecord(preTransformRecord);
+            final SourceRecord record = transformationChain.apply(preTransformRecord, operationExecutor, processingContext);
 
             if (record == null) {
                 counter.skipRecord();
                 commitTaskRecord(preTransformRecord);
                 continue;
             }
 
-            RecordHeaders headers = convertHeaderFor(record);
-            byte[] key = keyConverter.fromConnectData(record.topic(), record.keySchema(), record.key());
-            byte[] value = valueConverter.fromConnectData(record.topic(), record.valueSchema(), record.value());
+            RecordHeaders headers;
+            processingContext.setStage(Stage.HEADER_CONVERTER, headerConverter.getClass());
+            Result<RecordHeaders> headersResult = execute(() -> convertHeaderFor(record));
```

Comment:
Looks like setting the stage, processing the result can all be handled in the execute method